File: deepeval/metric_templates/resolver.py
from __future__ import annotations

import json
from importlib import resources
from pathlib import Path
from typing import Any, Dict, Optional, Set, Tuple

import jinja2


class MetricTemplateNotFoundError(KeyError):
    def __init__(self, message: str) -> None:
        super().__init__(message)
        self.message = message


class MetricTemplateInterpolationError(ValueError):
    def __init__(self, message: str, unresolved: Set[str]) -> None:
        super().__init__(message)
        self.unresolved = unresolved


class TemplateRegistry:
    """Loads the shipped base (English) `templates.json` and caches the Jinja
    environments plus compiled templates. Templates are static, so each
    `(class, method, strict)` is compiled at most once."""

    def __init__(self) -> None:
        self._base_templates: Optional[Dict[str, Any]] = None
        self._jinja_strict: Optional[jinja2.Environment] = None
        self._jinja_lenient: Optional[jinja2.Environment] = None
        self._compiled: Dict[Tuple[str, str, bool], jinja2.Template] = {}

    def clear(self) -> None:
        self._base_templates = None
        self._jinja_strict = None
        self._jinja_lenient = None
        self._compiled.clear()

    def get_base_templates(self) -> Dict[str, Any]:
        if self._base_templates is None:
            try:
                ref = resources.files("deepeval.metric_templates").joinpath(
                    "templates.json"
                )
                raw = ref.read_text(encoding="utf-8")
            except (ModuleNotFoundError, OSError, TypeError, FileNotFoundError):
                raw = (
                    Path(__file__).resolve().parent / "templates.json"
                ).read_text(encoding="utf-8")
            self._base_templates = json.loads(raw)
        return self._base_templates

    def get_jinja_env(self, strict: bool) -> jinja2.Environment:
        if strict:
            if self._jinja_strict is None:
                self._jinja_strict = jinja2.Environment(
                    undefined=jinja2.StrictUndefined
                )
            return self._jinja_strict
        if self._jinja_lenient is None:
            self._jinja_lenient = jinja2.Environment()
        return self._jinja_lenient

    def get_compiled(
        self, class_name: str, method: str, strict: bool
    ) -> jinja2.Template:
        key = (class_name, method, strict)
        template = self._compiled.get(key)
        if template is None:
            raw = get_raw_template(class_name, method)
            template = self.get_jinja_env(strict).from_string(raw)
            self._compiled[key] = template
        return template


_registry = TemplateRegistry()


def clear_metric_template_cache() -> None:
    _registry.clear()


def get_raw_template(class_name: str, method: str) -> str:
    """Return the raw (un-rendered) base template string for a class/method."""
    base = _registry.get_base_templates()
    entry = base.get(class_name, {})
    body = entry.get(method)

    if not body or not isinstance(body, str):
        raise MetricTemplateNotFoundError(
            f"No template found for {class_name!r}.{method!r}. "
            f"Available classes: {', '.join(base.keys())}"
        )
    return body


def iter_base_template_methods(class_name: str) -> list[tuple[str, str]]:
    """Iterate over all `(method, template)` pairs for a class in the base bundle."""
    base = _registry.get_base_templates()
    entry = base.get(class_name)

    if not entry or not isinstance(entry, dict):
        raise MetricTemplateNotFoundError(
            f"No metric templates for class {class_name!r}."
        )

    pairs = [
        (k, v)
        for k, v in sorted(entry.items())
        if not k.startswith("_") and isinstance(v, str)
    ]
    if not pairs:
        raise MetricTemplateNotFoundError(
            f"No string template methods for class {class_name!r}."
        )
    return pairs


def resolve_template(
    class_name: str,
    method: str,
    *,
    multimodal: bool = False,
    strict: bool = True,
    **kwargs: Any,
) -> str:
    """Render a base metric template to a final prompt via Jinja2.

    `_fragments` (shared reusable snippets) and `multimodal` are always
    available to the template; everything else is passed through `kwargs`.
    """
    fragments = _registry.get_base_templates().get("_fragments", {})
    try:
        template = _registry.get_compiled(class_name, method, strict)
        return template.render(
            multimodal=multimodal,
            _fragments=fragments,
            **kwargs,
        )
    except jinja2.UndefinedError as e:
        raise MetricTemplateInterpolationError(
            f"Missing variable during template render: {e.message}",
            unresolved=set(),
        ) from e
    except jinja2.TemplateSyntaxError as e:
        raise MetricTemplateInterpolationError(
            f"Jinja syntax error in template: {e.message}",
            unresolved=set(),
        ) from e

File: typescript/src/templates/resolver.ts
import nunjucks from "nunjucks";
// Per-feature bundles, mirroring deepeval/templates/<feature>/templates.json.
// scripts/compile_metric_templates.py emits these into BOTH packages.
import metricsBundle from "./metrics/templates.json";
const FRAGMENTS_KEY = "_fragments";

type TemplateBundle = Record<string, Record<string, string>>;
const BUNDLES: Record<string, TemplateBundle> = {
  metrics: metricsBundle as unknown as TemplateBundle,
};

function getBundle(feature: string): TemplateBundle {
  const b = BUNDLES[feature];
  if (!b) {
    throw new MetricTemplateNotFoundError(
      `No template bundle for feature '${feature}'. ` +
        `Available: ${Object.keys(BUNDLES).join(", ")}`,
    );
  }
  return b;
}

// --- Jinja2 parity helpers -------------------------------------------------
// Nunjucks and Jinja2 differ in two output details; we normalize the TS side to
// match Python so the SAME template renders byte-identically in both engines:
//   1. Arrays: Jinja prints a list as Python `repr` (`['a', 'b']`); Nunjucks
//      joins with commas (`a,b`). We give array values a Python-style toString
//      while leaving them real arrays (so `list[0]` / `{% for %}` still work).
//   2. Trailing newline: Jinja's default `keep_trailing_newline=False` drops one
//      trailing "\n"; Nunjucks keeps it. We strip one at compile time.

/** Python `repr()` of a value, as it appears INSIDE a list/dict (strings quoted). */
function pyRepr(v: unknown): string {
  if (v === null || v === undefined) return "None";
  if (typeof v === "string") {
    return (
      "'" +
      v
        .replace(/\\/g, "\\\\")
        .replace(/'/g, "\\'")
        .replace(/\n/g, "\\n")
        .replace(/\r/g, "\\r")
        .replace(/\t/g, "\\t") +
      "'"
    );
  }
  if (typeof v === "boolean") return v ? "True" : "False";
  if (typeof v === "number") return String(v);
  if (Array.isArray(v)) return "[" + v.map(pyRepr).join(", ") + "]";
  if (typeof v === "object") {
    return (
      "{" +
      Object.entries(v as Record<string, unknown>)
        .map(([k, val]) => `${pyRepr(k)}: ${pyRepr(val)}`)
        .join(", ") +
      "}"
    );
  }
  return String(v);
}

/**
 * Recursively give arrays a Python-`repr` `toString` so `{{ list }}` matches
 * Jinja, without losing array behavior (indexing, iteration). Non-arrays pass
 * through unchanged so booleans/null stay usable in `{% if %}` and objects keep
 * property access (e.g. `_fragments.x`, `obj.attr`).
 */
function toJinjaParityValue(v: unknown): unknown {
  if (Array.isArray(v)) {
    const arr = v.map(toJinjaParityValue);
    Object.defineProperty(arr, "toString", {
      value: () => pyRepr(arr),
      enumerable: false,
      configurable: true,
    });
    return arr;
  }
  return v;
}

export class MetricTemplateNotFoundError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "MetricTemplateNotFoundError";
  }
}

export class MetricTemplateInterpolationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "MetricTemplateInterpolationError";
  }
}

// --- cached environments + compiled templates (templates are static) ---
let _envStrict: nunjucks.Environment | null = null;
let _envLenient: nunjucks.Environment | null = null;
const _compiled = new Map<string, nunjucks.Template>();

function getEnv(strict: boolean): nunjucks.Environment {
  // autoescape MUST be off — these are LLM prompts, not HTML (matches the
  // Python jinja2.Environment default). `throwOnUndefined` mirrors StrictUndefined.
  if (strict) {
    if (_envStrict === null) {
      _envStrict = new nunjucks.Environment(null, {
        autoescape: false,
        throwOnUndefined: true,
      });
    }
    return _envStrict;
  }
  if (_envLenient === null) {
    _envLenient = new nunjucks.Environment(null, {
      autoescape: false,
      throwOnUndefined: false,
    });
  }
  return _envLenient;
}

/** Return the raw (un-rendered) template string for a class/method. */
export function getRawTemplate(
  feature: string,
  className: string,
  method: string,
): string {
  const bundle = getBundle(feature);
  const body = bundle[className]?.[method];
  if (typeof body !== "string" || body.length === 0) {
    throw new MetricTemplateNotFoundError(
      `No template found for '${feature}'.'${className}'.'${method}'. ` +
        `Available classes: ${Object.keys(bundle).join(", ")}`,
    );
  }
  return body;
}

function getCompiled(
  feature: string,
  className: string,
  method: string,
  strict: boolean,
): nunjucks.Template {
  const key = `${feature} ${className} ${method} ${strict}`;
  let template = _compiled.get(key);
  if (template === undefined) {
    // Mirror Jinja's default `keep_trailing_newline=False`: drop one trailing newline.
    let raw = getRawTemplate(feature, className, method);
    if (raw.endsWith("\n")) raw = raw.slice(0, -1);
    template = nunjucks.compile(raw, getEnv(strict));
    _compiled.set(key, template);
  }
  return template;
}

export interface ResolveTemplateOptions {
  /** Throw on undefined variables (mirrors Python's StrictUndefined). Default true. */
  strict?: boolean;
}

/**
 * Render a metric template to a final prompt.
 *
 * `multimodal` (default false) and `_fragments` (shared snippets) are always
 * available to the template; everything else is supplied via `vars`.
 *
 * @example resolveTemplate("metrics", "AnswerRelevancyMetric", "generate_verdicts", { input, statements })
 */
export function resolveTemplate(
  feature: string,
  className: string,
  method: string,
  vars: Record<string, unknown> = {},
  opts: ResolveTemplateOptions = {},
): string {
  const strict = opts.strict ?? true;
  const fragments = getBundle(feature)[FRAGMENTS_KEY] ?? {};
  const context: Record<string, unknown> = { multimodal: false };
  for (const [k, v] of Object.entries(vars)) {
    context[k] = toJinjaParityValue(v);
  }
  context._fragments = fragments;
  try {
    return getCompiled(feature, className, method, strict).render(context);
  } catch (e) {
    if (e instanceof MetricTemplateNotFoundError) throw e;
    throw new MetricTemplateInterpolationError(
      `Failed to render template '${feature}'.'${className}'.'${method}': ${(e as Error).message}`,
    );
  }
}

/** Clear cached environments + compiled templates (mainly for tests). */
export function clearMetricTemplateCache(): void {
  _envStrict = null;
  _envLenient = null;
  _compiled.clear();
}
